polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+    // but should probably generate this var better.
+    let sym = var.value().as_symbol().unwrap();
+    let field_str = field.value().as_string().unwrap();
+    let new_var = Symbol::new(&format!("{}_dot_{}", sym.0, field_str));
+
+    // Record the relationship between the vars.
+    var_info
+        .field_relationships
+        .push((sym.clone(), field_str.to_string(), new_var.clone()));
+    new_var
+}
+
+fn process_exp(var_info: &mut VarInfo, exp: &Operation) -> Option<Term> {
+    match exp.operator {
+        Operator::And => {
+            for arg in &exp.args {
+                let inner_exp = arg.value().as_expression().unwrap();
+                process_exp(var_info, inner_exp);
+            }
+        }
+        Operator::Dot => {
+            // Dot operations return a var that can be unified with.
+            // We create a new var to represent the result of the operation.
+            assert_eq!(exp.args.len(), 2);
+            let mut var = exp.args[0].clone();
+            if let Ok(inner_exp) = var.value().as_expression() {
+                if inner_exp.operator != Operator::Dot {
+                    unimplemented!("Operations other than dot nested within a dot are not yet supported for data filtering.")
+                }
+                var = process_exp(var_info, inner_exp).unwrap();
+            }
+            let field = &exp.args[1];
+            let new_var = dot_var(var_info, var, field);
+            // Return the var so we can unify with it.
+            return Some(Term::new_temporary(Value::Variable(new_var)));
+        }
+        Operator::Isa => {
+            assert_eq!(exp.args.len(), 2);
+            let lhs = &exp.args[0];
+            let rhs = &exp.args[1];
+            if let Value::Pattern(Pattern::Instance(InstanceLiteral { tag, fields })) = rhs.value()
+            {
+                if !fields.fields.is_empty() {
+                    unimplemented!("Specializer fields are not yet supported for data filtering.")
+                }
+                assert!(fields.fields.is_empty());
+                let var = match lhs.value() {
+                    Value::Variable(var) | Value::RestVariable(var) => var.clone(),
+                    Value::Expression(op) if op.operator == Operator::Dot => {
+                        dot_var(var_info, op.args[0].clone(), &op.args[1])
+                    }
+                    _ => todo!(),
+                };
+                var_info.types.push((var, tag.clone().0))
+            } else {
+                unimplemented!("Non pattern specializers are not yet supported for data filtering.")
+            }
+        }
+        Operator::Unify | Operator::Eq | Operator::Assign => {
+            assert_eq!(exp.args.len(), 2);
+
+            let mut lhs = exp.args[0].clone();
+            if let Value::Expression(op) = lhs.value() {
+                lhs = process_exp(var_info, op).unwrap();
```

Comment:
This means we expect lhs & rhs to be Dot, right?

Could we break `process_exp` into `process_exp_dot`, etc. and only have `process_exp_dot` return a `Term`?

Reply:
Maybe, I'm not sure yet if anything else will have to pass out a term. Will put on my list of cleanups for later if not.

Reply:
Sg